structure/tree/avl.go

```diff
@@ -11,10 +11,11 @@ type AVL[T constraints.Ordered] struct {
 
 // NewAVL create a novel AVL tree
 func NewAVL[T constraints.Ordered]() *AVL[T] {
+	var leaf *Node[T]
```

Comment:
I don't think this was a problem to begin with. I'm curious to know why you think this was a bug. Then I would present my case for why I don't think (at the moment - might change my mind after seeing your argument) 😄

Reply:
You are right. For AVL and BinarySearch `this.NIL == nil`. There is no need to introduce a variable `leaf` which is equivalent to `nil` .

Reply:
Actually, the bug exists in `RB`. I didn't make the parent of sentinel value NIL pointing to NIL. As a result, I had to introduce a lot of nil's, which is misleading. 
In this commit, I replace all nil to `t.NIL`(In BinarySearch and AVL, it's equal to nil. In RB, it's equal to `leaf`, a sentinel value).

```go
// NewBinarySearch create a novel Binary-Search tree
func NewBinarySearch[T constraints.Ordered]() *BinarySearch[T] {
	return &BinarySearch[T]{
		binaryTree: &binaryTree[T]{
			Root: nil,
			NIL:  nil,
		},
	}
}

// NewAVL create a novel AVL tree
func NewAVL[T constraints.Ordered]() *AVL[T] {
	return &AVL[T]{
		binaryTree: &binaryTree[T]{
			Root: nil,
			NIL:  nil,
		},
	}
}

// Create a new Red-Black Tree
func NewRB[T constraints.Ordered]() *RB[T] {
	leaf := &Node[T]{Color: Black, Left: nil, Right: nil}
	leaf.Parent = leaf  // the parent of NIL is pointing to NIL now
	return &RB[T]{
		binaryTree: &binaryTree[T]{
			Root: leaf,
			NIL:  leaf,
		},
	}
}

```

Reply:
Doesn't this mean that the bug where it exists should be solved instead of changing everything that is dependent on a base that is shared between different implementations. What you're currently doing is that you're fixing a bug in one implementation (RB) that are then causing issues in other implementations elsewhere. In real productions systems, this should be avoided at all costs otherwise this can easily become an issue (in this case its a small issue but trust me, it can create further issues in big systems).